cloudigrade/account/util.py

```diff
@@ -126,80 +135,134 @@ def generate_aws_ami_messages(instances_data, ami_list):
-        tuple(kombu.Exchange, kombu.Queue)
+        str: the queue's URL.
 
     """
-    exchange = kombu.Exchange(
-        settings.QUEUE_EXCHANGE_NAME,
-        'direct',
-        durable=True
-    )
-    message_queue = kombu.Queue(
-        queue_name,
-        exchange=exchange,
-        routing_key=queue_name
+    sqs = boto3.client('sqs')
+    try:
+        return sqs.get_queue_url(QueueName=queue_name)['QueueUrl']
+    except ClientError as e:
+        if e.response['Error']['Code'].endswith('.NonExistentQueue'):
+            return sqs.create_queue(QueueName=queue_name)['QueueUrl']
+        raise
+
+
+def _sqs_wrap_message(message):
+    """
+    Wrap the message in a dict for SQS batch sending.
+
+    Args:
+        message (object): message to encode and wrap
+
+    Returns:
+        dict: structured entry for sending to send_message_batch
+
+    """
+    return {
+        'Id': str(uuid.uuid4()),
+        # Yes, the outgoing message uses MessageBody, not Body.
+        'MessageBody': base64.b64encode(
+            jsonpickle.encode(message).encode('utf-8')
+        ).decode('utf-8'),
+    }
+
+
+def _sqs_unwrap_message(sqs_message):
+    """
+    Unwrap the sqs_message to get the original message.
+
+    Args:
+        sqs_message (dict): object to unwrap and decode
+
+    Returns:
+        object: the unwrapped and decoded message object
+
+    """
+    return jsonpickle.decode(
+        base64.b64decode(
+            # Yes, the response has Body, not MessageBody.
+            sqs_message['Body'].encode('utf-8')
+        ).decode('utf-8')
     )
-    return exchange, message_queue
 
 
 def add_messages_to_queue(queue_name, messages):
     """
-    Send messages to a message queue.
+    Send messages to an SQS queue.
 
     Args:
         queue_name (str): The queue to add messages to
         messages (list[dict]): A list of message dictionaries. The message
             dicts will be serialized as JSON strings.
-
-    Returns:
-        dict: A mapping of machine image ID to result boolean value
-
     """
-    exchange, message_queue = _create_exchange_and_queue(queue_name)
+    queue_url = _get_sqs_queue_url(queue_name)
+    sqs = boto3.client('sqs')
 
-    with kombu.Connection(settings.CELERY_BROKER_URL) as conn:
-        producer = conn.Producer(serializer='json')
-        for message in messages:
-            producer.publish(
-                message,
-                retry=True,
-                exchange=exchange,
-                routing_key=queue_name,
-                declare=[message_queue]
-            )
+    wrapped_messages = [_sqs_wrap_message(message) for message in messages]
+    batch_count = math.ceil(len(messages) / SQS_SEND_BATCH_SIZE)
+
+    for batch_num in range(batch_count):
+        start_pos = batch_num * SQS_SEND_BATCH_SIZE
+        end_pos = start_pos + SQS_SEND_BATCH_SIZE - 1
+        batch = wrapped_messages[start_pos:end_pos]
+        sqs.send_message_batch(QueueUrl=queue_url, Entries=batch)
```

Comment:
Should we log failures?  It looks like boto provides failures in the response body if they occur.

Reply:
At this point, if we can't write to the queue, there must be some rather catastrophic failure, and I think it makes sense to let the exception raise to the top (where it should be logged anyway).

The only reason I am explicitly handling and _working around_ exceptions when **reading** from the queue is that if we have an exception on the `n`th message, we would completely lose the `< n` messages since they would have already been deleted but not yet passed into the following function/task/whatever. It seemed more necessary to me to gracefully recover in that case, even though in that case it would also be pretty unexpected and catastrophic for the queue to be raising errors.

Reply:
When an exception is raised to the top and logged, do we have the context to know what we were trying to do?  I haven't played with SQS a ton, but I know Kafka and other messaging queues can have issues where the system is up not not fully working.  Kafka can have a topic corrupted while another is working.  My suggestion for logging was to ensure we know if there is a global outage verse a localized failure (just a single queue).